### PR TITLE
Adding tests to the library

### PR DESCRIPTION
**Pull request by KristenGarnier**
Hi there,

I have added some tests to secure feature adding or refactoring in the future.

I had to do some minor modifications in order to get tests on some part of the system.
But it is not complete yet, i cloud only get 55% coverage before doing big changes in the library ( Isolate anounymous functions into their own method , otherwise it is not testable)

If you want i can do these changes, but i wanted to check with you first if that was possible and under which constraints.

Let me know about it,

Have a nice day !

### CODE REVIEW

**Comment by KristenGarnier** [comment]
Might also be a good thing to setup test coverage in codacy

**Comment by KristenGarnier** [comment]
Might also be a good thing to setup test coverage in codacy

**Comment by ColinEspinas** [comment]
Hey,
Do all the changes you want to get full coverage, we'll see how it goes and change what needs to change once it is done.
Also, why are you using a `__` prefix for your functions ? To isolate them from the API functions ?

Have a nice day too.

**Comment by KristenGarnier** [comment]
Hi,

I use "__" to isolate internal method from public API, since javascript has no private built in.

I will try to hide them differently once all test has been done in another pull request if i have time

**Comment by ColinEspinas** [comment]
No problem, just try to keep the code clean and not to increase the size of the library too much.

Also I encourage you to use the [conventional commits specifications](https://www.conventionalcommits.org/en/v1.0.0/) for your commit on darken, it is not a requirement but I recommend it.

**Comment by KristenGarnier** [comment]
I am done here,
up to 92% test coverage, only function that listens on match media are not tested

**Comment by ColinEspinas** [comment]
Nice thanks for the help, a last question, why are you adding babel to jest ?

**Comment by KristenGarnier** [comment]
I do this for the export statement, node does not support that out of the box (uses CommonJs require)

It also help to even out the differences between versions of NodeJs, so i can just use es6+ code without worrying about it

**Comment by ColinEspinas** [comment]
I see, I use microbundle to handle all the es6+ and exports maybe there could be a way to make it work with it instead of adding babel ?

**Comment by KristenGarnier** [comment]
Hi,

While this is possible, this is something i would not recommend doing, here is why: 
- Adding a compilation step at every change, when user just want the test to run in watch mode
- Do not allow user to test by file : Let's say you have a second file, you cannot target test directly to this, but only to the production build
- Limit the debugging capabilities : You can do less in class debugging to check direct comportments

TLDR : Using microbundle for compiler limit the flexibility of the testing environment

If you want microbundle i will make it work with microbundle, just consider some of tradeoffs before you post the answer. 

## Note about size

If you worry about size of the bundle, microbundle will not bundle babel or tests into the production files, i checked the files size after compilation : 

```sh
[With tests] Build "darken" to dist:
        750 B: darken.js.gz
        626 B: darken.js.br
        748 B: darken.mjs.gz
        621 B: darken.mjs.br
        802 B: darken.umd.js.gz
        675 B: darken.umd.js.br

[Master] Build "darken" to dist:
        686 B: darken.js.gz
        564 B: darken.js.br
        682 B: darken.mjs.gz
        556 B: darken.mjs.br
        741 B: darken.umd.js.gz
        608 B: darken.umd.js.br
```

The notable size difference is the refactoring, since some repeating code have been changed to one general method, gzip and brotli cannot it optimize it as well as before which yields a 10% increase in compressed sizes. ( That would not change if i use microbundle or babel directly)
